tests/src/metadata.rs

```diff
@@ -0,0 +1,257 @@
+    pub compare_ref: Option<bool>,
+    /// Hint annotations will be compared to compiler hints.
+    ///
+    /// Defaults to `true`, can be disabled with `Hints: false`.
+    pub validate_hints: Option<bool>,
+    /// Autocompletion annotations will be validated against autocompletions.
+    /// Mutually exclusive with error and hint annotations.
+    ///
+    /// Defaults to `false`, can be enabled with `Autocomplete: true`.
+    pub validate_autocomplete: Option<bool>,
+}
+
+/// Annotation of the form `// KIND: RANGE TEXT`.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct Annotation {
+    /// Which kind of annotation this is.
+    pub kind: AnnotationKind,
+    /// May be written as:
+    /// - `{line}:{col}-{line}:{col}`, e.g. `0:4-0:6`.
+    /// - `{col}-{col}`, e.g. `4-6`:
+    ///    The line is assumed to be the line after the annotation.
+    /// - `-1`: Produces a range of length zero at the end of the next line.
+    ///   Mostly useful for autocompletion tests which require an index.
+    pub range: Option<Range<usize>>,
+    /// The raw text after the annotation.
+    pub text: EcoString,
+}
+
+/// The different kinds of in-test annotations.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum AnnotationKind {
+    Error,
+    Warning,
+    Hint,
+    AutocompleteContains,
+    AutocompleteExcludes,
+}
+
+impl AnnotationKind {
+    /// Returns the user-facing string for this annotation.
+    pub fn as_str(self) -> &'static str {
+        match self {
+            AnnotationKind::Error => "Error",
+            AnnotationKind::Warning => "Warning",
+            AnnotationKind::Hint => "Hint",
+            AnnotationKind::AutocompleteContains => "Autocomplete contains",
+            AnnotationKind::AutocompleteExcludes => "Autocomplete excludes",
+        }
+    }
+}
+
+impl FromStr for AnnotationKind {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "Error" => AnnotationKind::Error,
+            "Warning" => AnnotationKind::Warning,
+            "Hint" => AnnotationKind::Hint,
+            "Autocomplete contains" => AnnotationKind::AutocompleteContains,
+            "Autocomplete excludes" => AnnotationKind::AutocompleteExcludes,
+            _ => return Err("invalid annotatino"),
+        })
+    }
+}
+
+impl Display for AnnotationKind {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.pad(self.as_str())
+    }
+}
+
+/// Parse metadata for a test.
+pub fn parse_part_metadata(source: &Source) -> TestMetadata {
+    let mut config = TestConfig::default();
+    let mut annotations = HashSet::default();
+    let mut invalid_data = vec![];
+
+    let lines: Vec<_> = source.text().lines().map(str::trim).collect();
+    let comments_until_code =
+        |i: usize| lines[i..].iter().take_while(|line| line.starts_with("//")).count();
+
+    let pos = |s: &mut Scanner, i: usize| -> Option<usize> {
```

Comment:
Ideally, these two closures should be extracted. They'll need an extra `source` argument though to do the conversion and comment handling.